### PR TITLE
Grid-based particle initialization and random init rewrite

### PR DESCRIPTION
**Pull request by streeve**
- Both uniform and random creation from Cajita grids
- Updates the core library init to make it more similar (type tags and non-uniform box). Behavior changes to ignore particles for the variants with minimum particle separation (and therefore may not return the number of particles requested)
- Deprecates previous simple random init versions (these still loop until all particles have been created which is extremely slow)

~~Core versions should also accept `ParticleList` + user init functors~~ done

### CODE REVIEW

**Comment by streeve** [comment]
@abisner this is finally ready for review

**Comment by streeve** [comment]
Now blocked by #654

**Comment by streeve** [comment]
@abisner this actually passes all tests. I removed the filter empty routine for now because of SYCL issues, but I think it's best to revisit that later

File: cajita/src/Cajita_ParticleInit.hpp [hunk 1]
```diff
@@ -0,0 +1,586 @@
...
+ *                                                                          *
+ * This file is part of the Cabana library. Cabana is distributed under a   *
+ * BSD 3-clause license. For the licensing terms see the LICENSE file in    *
+ * the top-level directory.                                                 *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Copyright (c) 2021 by the Picasso authors                                *
```
Comment (lebuller) [inline cajita/src/Cajita_ParticleInit.hpp:13]: Should the picasso copyright header be left in here?

**Comment by streeve** [comment]
Failures unrelated

**Pull request merged**